Bugfix: Package relay / bytespersigop checks

luke-jr opened the pull request:
`GetVirtualTransactionSize(CTransaction&)` passes by default 0 sigops and 0 `-bytespersigop`, which is incorrect. In many cases (ie, wallet), this is harmless since we have control over the transactions and don't do anything absurd in normal usage. But in other cases (including wallet *received* transactions), this isn't safe. To avoid invisible bugs like this, I delete (or rather, move to `bitcoin_test`) the function signature that allows for omitting sigop inputs.

This also then fixes the new package relay code to use the correct virtual sizes for its checks, and documents calling locations where the behaviour is broken or safe.

In particular, note that the `sendrawtransaction` RPC and GUI transaction details remain buggy and not fixed in this PR.

luke-jr committed:
policy: Delete buggy GetVirtualTransactionSize(CTransaction&)

luke-jr commented:
See also https://github.com/bitcoin-core/gui/issues/750

glozow commented on src/txmempool.cpp:
```diff
@@ -197,13 +197,15 @@ util::Result<CTxMemPool::setEntries> CTxMemPool::CalculateAncestorsAndCheckLimit
 }
 
 bool CTxMemPool::CheckPackageLimits(const Package& package,
+                                    const std::vector<int64_t> &all_tx_vsizes,
```
in fef601e2a4d790781fee316c1a55ea48a134a0f1
Given that this function is only interested in the total size, why not just pass in a `int64_t total_vsize`?

instagibbs replied:
> why not just pass in a int64_t total_vsize?

Indeed, then you don't have to gather all the vsizes twice as well.

luke-jr replied:
Done

glozow commented on src/validation.cpp:
```diff
@@ -987,6 +988,7 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
 }
 
 bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txns,
+                                         const std::vector<int64_t> &all_tx_vsizes,
```
in fef601e2a4d790781fee316c1a55ea48a134a0f1
nit: perhaps pass the vector of `workspaces` instead? It would follow the pattern of the other `MemPoolAccept` functions and be more extendable for future stuff.

glozow commented on src/validation.cpp:
```diff
@@ -1277,6 +1277,10 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
     const auto m_total_vsize = std::accumulate(workspaces.cbegin(), workspaces.cend(), int64_t{0},
         [](int64_t sum, auto& ws) { return sum + ws.m_vsize; });
+    if (m_total_vsize > MAX_PACKAGE_SIZE * 1000) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-too-large");
```
in 9307e5c6fc1786e5fc734da1685e1cde52f4dd61

good catch, perhaps an error string like `package-exceeds-max-vsize` would help distinguish the two errors.

luke-jr replied:
Should they be distinguished?

glozow replied:
I think these are distinct rules (where "vsize" means sigop-adjusted):
(a) Each tx has a maximum vsize
(b) Ancestor and descendant packages within the mempool cannot exceed 101kvB vsize
(c) "Packages cannot exceed `MAX_PACKAGE_SIZE=101KvB` total size" ([doc](https://raw.githubusercontent.com/bitcoin/bitcoin/master/doc/policy/packages.md)).

The bug being fixed here is that we don't properly enforce (b) when we're validating packages.
I think your commit "Bugfix: Pass correct virtual size to CheckPackageLimits" can be followed by something like [this commit](https://github.com/bitcoin/bitcoin/pull/28450/commits/ad63201bd3c4b41395278f4b45934727471bacdf) instead of a new "package-too-large" check.

Imo the intent of (c) is to make sure we don't work on packages that are too large / potentially expensive to validate. For that, I think we should just change the description of (c) to "Packages cannot exceed `MAX_PACKAGE_SIZE=404kWu` total weight."

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -84,7 +84,7 @@ static RPCHelpMan sendrawtransaction()
                                                      CFeeRate(AmountFromValue(request.params[1]));
 
-            int64_t virtual_size = GetVirtualTransactionSize(*tx);
+            int64_t virtual_size = GetVirtualTransactionSize(*tx, 0, 0 /* BUG */);
```
Perhaps the way to fix this is to overload `BroadcastTransaction()` to accept either a `CFeeRate`, `CAmount`, or nothing. After the test-accept of the transaction, you can query `MempoolAcceptResult::m_vsize` to check the feerate.

luke-jr replied:
@glozow Isn't that exactly the behaviour that we refactored out of (to checking locally here)?

luke-jr replied:
Ah, I see what you mean. Fixed.

glozow reviewed (commented):
concept ACK

instagibbs commented on src/policy/policy.h:
```diff
@@ -158,11 +158,6 @@ int64_t GetVirtualTransactionSize(int64_t nWeight, int64_t nSigOpCost, unsigned
 int64_t GetVirtualTransactionSize(const CTransaction& tx, int64_t nSigOpCost, unsigned int bytes_per_sigop);
 int64_t GetVirtualTransactionInputSize(const CTxIn& tx, int64_t nSigOpCost, unsigned int bytes_per_sigop);
 
-static inline int64_t GetVirtualTransactionSize(const CTransaction& tx)
```
Alternative, move to wallet and rename to `GetWalletVirtualTransactionSize` with tests/assertions/comments saying why it's safe in one location instead of spread around? Would make grepping for the remaining bugs easier as well. e.g., anything like `GetVirtualTransactionSize(tx, 0, 0)` would be obviously a bug at that point

instagibbs commented on src/qt/walletmodeltransaction.h:
```diff
@@ -29,6 +29,7 @@ class WalletModelTransaction
     CTransactionRef& getWtx();
     void setWtx(const CTransactionRef&);
 
+    //! WARNING: Currently only safe for self-generated transactions due to bypassing sigop logic of GetVirtualTransactionSize
```
seems a bit scary with no actionable suggestion?

luke-jr replied:
I don't see any obvious actionable solution (https://github.com/bitcoin-core/gui/issues/750). But this is currently only called where it is safe.

instagibbs reviewed (commented):
concept ACK

luke-jr commented on src/test/util/transaction_utils.h:
```diff
@@ -26,4 +27,10 @@ CMutableTransaction BuildSpendingTransaction(const CScript& scriptSig, const CSc
 // the second nValues[2] and nValues[3] outputs paid to a TxoutType::PUBKEYHASH.
 std::vector<CMutableTransaction> SetupDummyInputs(FillableSigningProvider& keystoreRet, CCoinsViewCache& coinsRet, const std::array<CAmount,4>& nValues);
 
+// Assumes a non-spammy tx
```
Done

luke-jr commented:
`checkChainLimits` should probably get a warning-only-for-wallet-use comment too

luke-jr committed:
node: Refactor BroadcastTransaction to accept a CFeeRate maximum

luke-jr committed:
Bugfix: RPC/Mempool: Pass CFeeRate to BroadcastTransaction to correctly account for non-weight vsize

luke-jr committed:
Bugfix: GUI/PSBTOperationsDialog: Pass feerate limit as a CFeeRate instead of assuming 1kvB

instagibbs commented:
could we work on getting tests working? looking to do some deeper review

luke-jr commented:
The CI failures appear to have been because CFeeRate(0) needed to be special-cased as "no limit" also. Fixed that.

Rebased on top of #28471, partially by reverting its first commit which appears to simply make the bug harder to fix.

glozow commented:
@luke-jr I don’t think adding a new vsize-based total package limit is
necessary, the existing ancestor size limits are sufficient. #28471 already
incorporated your bug fix to have CheckPackageLimits look at package total
vsize when enforcing ancestor size limits. If you feel there’s an
additional bug, can you please write a test that fails on #28471 but passes
on this branch?

El El vie, 22 sept 2023 a las 4:05, Luke Dashjr ***@***.***>
escribió:

> The CI failures appear to have been because CFeeRate(0) needed to be
> special-cased as "no limit" also. Fixed that.
>
> Rebased on top of #28471 <https://github.com/bitcoin/bitcoin/pull/28471>,
> partially by reverting its first commit which appears to simply make the
> bug harder to fix.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/28345#issuecomment-1730772463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGAEGGMJ5MNIMSGJ2OD4IVTX3UFBZANCNFSM6AAAAAA3727CQI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

luke-jr commented:
> I don’t think adding a new vsize-based total package limit is
necessary, the existing ancestor size limits are sufficient.

Ok, dropping that (and the revert) then.

glozow reviewed (commented):
LGTM though agree with @instagibbs' suggestion for the first commit. Could update the title + description? No longer package relay-related.

ismaelsadeeq reviewed (commented):
Concept ACK

achow101 commented:
Closing as up for grabs due to lack of activity.
